hal/src/sercom/v2/i2c.rs

```diff
@@ -0,0 +1,426 @@
+//! type Config = i2c::Config<Pads>;
+//!
+//! let pm = peripherals.PM;
+//! let sercom = peripherals.SERCOM0;
+//! // Configure GCLK for 10 MHz
+//! let freq = 10.mhz();
+//! let config = i2c::Config::new(&pm, sercom, pads, freq);
+//! ```
+//!
+//! The [`Config`] struct can configure the peripheral in one of two ways:
+//!
+//! * A set of methods is provided to use in a builder pattern: for example
+//!   [`baud`](Config::baud), [`run_in_standby`](Config::run_in_standby), etc.
+//!   These methods take `self` and return `Self`.
+//! * A set of methods is provided to use as setters: for example
+//!   [`set_baud`](Config::set_baud),
+//!   [`set_run_in_standby`](Config::set_run_in_standby), etc. These methods
+//!   take `&mut self` and return nothing.
+//!
+//! In any case, the peripheral setup ends with a call to [`enable`], which
+//! consumes the [`Config`] and returns an enabled [`I2c`] peripheral.
+//!
+//! ```
+//! let i2c = i2c::Config::new(&pm, sercom, pads, freq)
+//!     .baud(1.mhz())
+//!     .enable();
+//! ```
+//!
+//! Alternatively,
+//!
+//! ```
+//! let i2c = i2c::Config::new(&mclk, sercom, pads, freq);
+//!     i2c.set_baud(1.mhz());
+//!     let i2c = i2c.enable();
+//! ```
+//!
+//! ## Reading the current configuration
+//!
+//! It is possible to read the current configuration by using the getter methods
+//! provided: for example [`get_baud`](Config::get_baud),
+//! [`get_run_in_standby`](Config::get_run_in_standby), etc.
+//!
+//! # [`I2c`]
+//!
+//! [`I2c`] structs can only be created from a [`Config`]. They have one type
+//! parameter, representing the underlying [`Config`].
+//!
+//! Only the [`I2c`] struct can actually perform
+//! transactions. To do so, use the embedded HAL traits, like
+//! [`i2c::WriteRead`], [`i2c::Read`] and [`i2c::Write`].
+//!
+//! ```
+//! use embedded_hal::blocking::i2c::Write;
+//!
+//! i2c.write(0x54, 0x0fe)
+//! ```
+//!
+//! # Reading the current configuration
+//!
+//! The `AsRef<Config<P>>` trait is implemented for `I2c<Config<P>>`.
+//! This means you can use the `get_` methods implemented for `Config`, since
+//! they take an `&self` argument.
+//!
+//! ```
+//! // Assume i2c is a I2c<C<P>>
+//! let baud = i2c.as_ref().get_baud();
+//! ```
+//!
+//! # Reconfiguring
+//!
+//! The [`reconfigure`] method gives out an `&mut Config` reference, which can
+//! then use the `set_*` methods.
+//!
+//! ```
+//! use atsamd_hal::sercom::v2::i2c::I2c;
+//! use atsamd_hal::time::*;
+//!
+//! // Assume config is a valid Duplex I2C Config struct
+//! let i2c = config.enable();
+//!
+//! // Send/receive data...
+//!
+//! // Reconfigure I2C peripheral
+//! i2c.reconfigure(|c| c.set_run_in_standby(false));
+//!
+//! // Disable I2C peripheral
+//! let config = i2c.disable();
+//! ```
+//!
+//! # Non-supported features
+//!
+//! * Slave mode is not supported at this time.
+//! * High-speed mode is not supported.
+//! * 4-wire mode is not supported.
```

Comment:
4 wire I2C?? 😮

Reply:
Haha, I don't know what the hell it's supposed to be used for, but the datasheet says it's supported!